src/cryptonote_core/blockchain.cpp

```diff
@@ -3039,12 +3042,14 @@ bool Blockchain::check_tx_outputs(const transaction& tx, tx_verification_context
   // from v4, forbid invalid pubkeys
   if (hf_version >= 4) {
     for (const auto &o: tx.vout) {
-      if (o.target.type() == typeid(txout_to_key)) {
-        const txout_to_key& out_to_key = boost::get<txout_to_key>(o.target);
-        if (!crypto::check_key(out_to_key.key)) {
-          tvc.m_invalid_output = true;
-          return false;
-        }
+      crypto::public_key output_public_key;
+      if (!get_output_public_key(o, output_public_key)) {
+        tvc.m_invalid_output = true;
+        return false;
+      }
+      if (!crypto::check_key(output_public_key)) {
+        tvc.m_invalid_output = true;
+        return false;
       }
```

Comment:
Seems to me that the way you check now only allows for the two output types that you support, but not other ones that with the old code are just skipped if I read the code correctly. Is that so, are `txout_to_script` and `txout_to_scripthash` dead for good?

Reply:
That's correct.

This is worth double checking, but I scanned the entire chain to make sure there are no `txout_to_script` or `txout_to_scripthash` types in the chain already, and made sure they can't be added to the chain. There are 2 things in the existing code that prevent them from being added ([1](https://github.com/monero-project/monero/blob/a9cb4c082f1a815076674b543d93159a2137540e/src/cryptonote_basic/cryptonote_format_utils.cpp#L807), [2](https://github.com/monero-project/monero/blob/6d94953be259c6bf4a1407ec552a570a66b6179d/src/blockchain_db/lmdb/db_lmdb.cpp#L1048-L1049)).

I also had a similar hesitation on this as you, which is why I checked the above before making the change -- can see discussion on it with koe [here](https://github.com/monero-project/monero/pull/8061#discussion_r749644658) and [here](https://github.com/monero-project/monero/pull/8061#discussion_r749634449).